### ISSUE TITLE
[UI Builder] Stage free-transformation problem

### ISSUE THREAD

**Issue by rat-moonshine** [prominic/Moonshine-IDE#137]
![ccproblem3](https://user-images.githubusercontent.com/24370870/34823782-e8b6142a-f6f1-11e7-89e8-e4c6acbd979d.gif)

**Comment by piotrzarzycki21** [comment]
I just looked into that and I'm not so sure whether this a problem. The editing area is depends on screen which you have at some point. On my laptop where I have small resolution non HD I think so 1366x768 the editing area is small. 
Thanks to scrollbars I can create bigger window with more elements. 
The solution here is to remove scrollbars and than Editing area will be restricted, but I'm not in a favor of that resolution.

Resizing "Main Window" to the points where scrollbars appear is in my opinion ok. Maybe restricting resizing of components which are inside "Main Window" is something which I should look, but also I'm not so convinced to do that.

Thoughts ?

**Comment by rat-moonshine** [comment]
In my opinion, a few changes may improve the scenario:
1. The main editor stage - we can stop it from being resized like the way it is now (which creates some of these complexities)
2. Keep the editor stage scrollable (as was before)
3. Introduction of HDividerBox between the left/middle/right panel (or probably some mechanism to collapse/expand) so an user can get more space when requires

**Comment by piotrzarzycki21** [comment]
What do you mean by  "main editor stage" - It is the "Main Window" which we have introduced ? It could be non resizible by drag/drop as it is now. Resize could be done only through typing width/height - If that what's you mean in "1"

**Comment by rat-moonshine** [comment]
By 'main editor stage' I meant to the editor/stage where we can drag-n-drop UI components to. 

What do you meant by 'it could be non resizable by drag/drop as it is now' - if you also pointed to the area which I mentioned by 'main editor stage', you see its resizable (as shown in my screen capture).

**Comment by piotrzarzycki21** [comment]
Currently we have white main window with Title - This is our main area where we are drag/drop components. I can make this area non resizible by drag/drop.

**Comment by piotrzarzycki21** [comment]
'it could be non resizable by drag/drop as it is now' - By this I meant the ability to hold on the "dots" in the corner of the component and make resize.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine I just make MainApplication window non resizible by drag/drop. User still will be able to resize window by typing values in the text inputs. Take a look and let me know if it is something which you were talking about.

**Comment by rat-moonshine** [comment]
After having some more use, I came up with some thoughts:

- Initially I had the impression that the panel added to the stage by default (when project creation) - is the only stage where I should drop all my components (since its size). Which is wrong.
- I think we should get rid of adding any default panel to the editor during the project creation (because of the above confusion), and let user draw his/her own mockup from the scratch
- Since you stopped transformation to the default added panel - it is also became not-per-the-component-rule (I asked this to do because mistakenly I had the impression that its the only stage where user should drop all the components, but it is not!)

**Comment by piotrzarzycki21** [comment]
I will than get back to the previous behavior where you can resize Main Window (Main Panel). 

As for the default Panel - we have discussed this one with @JoelProminic in #98. This Panel is representation of Main Window, once you make an export to Flex application this Panel will be transformed into Main Window of exported application. 
Initially Title of that window was "Main Window", maybe we should get back to something more meaning full like: "Main application Window" instead of having project name there.

**Comment by JoelProminic** [comment]
I don't think we should get rid of the default panel, for the reasons I discussed in #98.   I also like being able to resize the window manually.

Having the scroll bars appear automatically makes sense to me, and the implementation seems to work well.  However, there are some visual oddities:

![image](https://user-images.githubusercontent.com/24532731/35058533-db2af7e0-fb7d-11e7-9754-a25cb8e9e44d.png)

You can see that the draggable points for the panel appear outside of the mockup window in my screenshot (red).  These do not change position when I scroll the mockup window.  Normally, I would expect the dragable point to appear in the mockup area as I scroll down.

Also, the division between the bottom bar and the mockup window is not clear, so it looks like the panel gets cut off (green).  If there was a divider or a difference in the background color, this would be more clear.

Having the subcomponents be cut off if they go outside of the base application panel makes sense to me, though it has a similar issue with the dragable points.

@rat-moonshine, is there some other major issue that I missed from your screen capture?

**Comment by rat-moonshine** [comment]
The problem in this design what I'm seeing are:
- If Piotr stopped those draggable points from showing when they're outside of the visible area - user may not able to resize back them again
- On the other hand the above design where draggable points are showing off-visible-area - is not right as well

I still strongly feel for many this such reasons that adding a default panel to the stage with its own fixed width/height is not right. To overcome #98 confusion we could have workaround by adding the background white color (maybe?) to make it more sense of a stage.

**Comment by piotrzarzycki21** [comment]
If I switch off resize option user can always dot resize manually. It is huge problem those points outside of the screen. I was fighting with them once in the issue where they appear in the settings project. I did fix that by unselecting component. In that case if we won't switch OFF resize of main window I'm not sure how to fix that yet.

**Comment by piotrzarzycki21** [comment]
I have looked into that again and see two solutions, which require lot's of work. Let me know if we want to do any of that. @rat-moonshine @JoelProminic 

1) Have option to open Visual editor Tab into separate window which prevents from the issue
2) Exchange dots buttons in the corner which indicates selection. Add ability to resize when user drag border of the window. Example:

![resize](https://user-images.githubusercontent.com/24554795/35219144-a8be195c-ff71-11e7-9516-85ea74793cf9.gif)

I think we should go with option two. Let me know whether you have some other ideas.

**Comment by piotrzarzycki21** [comment]
I won't spent time on that for 1.9.0 release. Will try to look into that for the next.

**Comment by JoelProminic** [comment]
I'm not clear what you are suggesting with your second option.   Your GIF only shows the sidebar panel for me - was it cut off?

**Comment by piotrzarzycki21** [comment]
@JoelProminic I'm showing just how I would change resizing for the components in Visual Editor. Currently you have 4 dots in each corner. I could work on a solution where you will be able to drag using border of the component, as it is shown on the GIF.
